labs/Evaluation/gradient_boosting.py

```diff
@@ -0,0 +1,180 @@
+            title=f"Gradient Boosting with max_depth={d}",
+            xlabel="nr estimators",
+            ylabel=metric,
+            percentage=True,
+        )
+    print(
+        f'GB best for {best_params["params"][2]} trees (d={best_params["params"][0]} and lr={best_params["params"][1]}'
+    )
+
+    return best_model, best_params
+
+def gradient_boosting_variables_studies(best_model, vars, path):
+    trees_importances: list[float] = []
+    for lst_trees in best_model.estimators_:
+        for tree in lst_trees:
+            trees_importances.append(tree.feature_importances_)
+
+    stdevs: list[float] = list(std(trees_importances, axis=0))
+    importances = best_model.feature_importances_
+    indices: list[int] = argsort(importances)[::-1]
+    elems: list[str] = []
+    imp_values: list[float] = []
+    for f in range(len(vars)):
+        elems += [vars[indices[f]]]
+        imp_values.append(importances[indices[f]])
+        print(f"{f+1}. {elems[f]} ({importances[indices[f]]})")
+
+    plt.figure(figsize=(9, 8))
+    plot_horizontal_bar_chart(
+        elems,
+        imp_values,
+        error=stdevs,
+        title="GB variables importance",
+        xlabel="importance",
+        ylabel="variables",
+        percentage=True,
+    )
+    if path:
+        plt.yticks(rotation=45)
+        savefig(path+"gradient_boosting_variables_study")
+
+def gradient_boosting_overfitting_study(trnX, trnY, eval_metric = "accuracy", path=False):
+    d_max: int = params["params"][0]
+    lr: float = params["params"][1]
+    nr_estimators: list[int] = [i for i in range(2, 2501, 500)]
+
+    y_tst_values: list[float] = []
+    y_trn_values: list[float] = []
+    acc_metric: str = "accuracy"
+
+    for n in nr_estimators:
+        clf = GradientBoostingClassifier(n_estimators=n, max_depth=d_max, learning_rate=lr)
+        clf.fit(trnX, trnY)
+        prd_tst_Y: array = clf.predict(tstX)
+        prd_trn_Y: array = clf.predict(trnX)
+        y_tst_values.append(CLASS_EVAL_METRICS[acc_metric](tstY, prd_tst_Y))
+        y_trn_values.append(CLASS_EVAL_METRICS[acc_metric](trnY, prd_trn_Y))
+
+    figure()
+    plot_multiline_chart(
+        nr_estimators,
+        {"Train": y_trn_values, "Test": y_tst_values},
+        title=f"GB overfitting study for d={d_max} and lr={lr}",
+        xlabel="nr_estimators",
+        ylabel=str(eval_metric),
+        percentage=True,
+    )
+    if path:
+        savefig(path+"gradient_boosting_overfitting_study")
+
+def study_and_train(train_filename, test_filename, target, path = False, eval_metric = "accuracy"):
+    trnX, tstX, trnY, tstY, labels, vars = read_train_test_from_files(
+        train_filename, test_filename, target
+    )
+    print(f"Train#={len(trnX)} Test#={len(tstX)}")
+    print(f"Labels={labels}")
+
+    figure()
+    best_model, params = gradient_boosting_study(
+        trnX,
+        trnY,
+        tstX,
+        tstY,
+        nr_max_trees=1000,
+        lag=250,
+        metric=eval_metric,
```

Comment:
Should we also run this for other metrics, such as recall and precision?

Reply:
Yeah, I was thinking the same 😅